nucliadb/nucliadb/common/cluster/manager.py

```diff
@@ -310,8 +316,53 @@ async def add_resource(
             indexpb.shard = replica_id
             await indexing.index(indexpb, node_id)
 
+    def should_create_new_shard(self, shard_info: noderesources_pb2.Shard) -> bool:
+        return (
+            shard_info.paragraphs > settings.max_shard_paragraphs
+            or shard_info.fields > settings.max_shard_fields
+        )
+
+    async def maybe_create_new_shard(
+        self, kbid: str, shard_info: noderesources_pb2.Shard
+    ):
+        if self.should_create_new_shard(shard_info):
+            logger.warning({"message": "Adding shard", "kbid": kbid})
+            kbdm = KnowledgeBoxDataManager(get_driver())
+            model = await kbdm.get_model_metadata(kbid)
+            driver = get_driver()
+            async with driver.transaction() as txn:
+                await self.create_shard_by_kbid(txn, kbid, semantic_model=model)
+                await txn.commit()
+
 
 class StandaloneKBShardManager(KBShardManager):
+    max_ops_before_checks = 25
+
+    def __init__(self):
+        super().__init__()
+        self._lock = asyncio.Lock()
+        self._change_count: dict[tuple[str, str], int] = {}  # type: ignore
+
+    async def _resource_change_event(
+        self, kbid: str, node_id: str, shard_id: str
+    ) -> None:
+        if (node_id, shard_id) not in self._change_count:
+            self._change_count[(node_id, shard_id)] = 0
+        self._change_count[(node_id, shard_id)] += 1
+        if self._change_count[(node_id, shard_id)] < self.max_ops_before_checks:
+            return
+
+        self._change_count[(node_id, shard_id)] = 0
+        async with self._lock:
+            index_node = get_index_node(node_id)
```

Comment:
IIRC, for standalone this will return a proxy index node and all the following method calls will be proxied to the `StandaloneClusterServiceStub` of each "standalone" node right?

Do you think it would be better to be more explicit with this? Either with the variable name `index_node` -> `proxy_index_node`, or with type hints `index_node: ProxyIndexNode`, or with a comment?

Reply:
I like it. added.